# Homework Assignment - Lesson 6: Combining Datasets - Joins

**Due Date:** 10/7/2025

**Instructions:**
- Complete the following tasks in this R notebook
- Use the exact variable names specified for each task
- Ensure your code runs without errors and produces the expected outputs
- Submit your completed notebook with all code and outputs

---

## Part 1: Data Import and Setup

**Task:** Import the following CSV files and examine their structure.

**Required Variable Names:**
You must use these EXACT variable names:
- `customers` for customers.csv
- `orders` for orders.csv  
- `order_items` for order_items.csv
- `products` for products.csv
- `suppliers` for suppliers.csv

**Expected Outputs:**
- All datasets loaded successfully
- Print dataset dimensions using `nrow()` and `ncol()`
- Show first 3 rows of each dataset using `head(dataset, 3)`

In [3]:
# Your code here:
# 1. Load the tidyverse package
# 2. Set working directory to sample_datasets  
# 3. Import all CSV files using read_csv() with EXACT variable names above
# 4. Print dimensions and head() for each dataset

# Load tidyverse
library(tidyverse)


# Set working directory
setwd("/workspaces/Carsons-stuff/data")
cat("✅ Working directory set to:", getwd(), "\n")

# Import datasets with exact variable names
customers <- read_csv("customers.csv")
orders <- read_csv("orders.csv")
order_items <- read_csv("order_items.csv")
products <- read_csv("products.csv")
suppliers <- read_csv("suppliers.csv")
cat("✅ Datasets imported successfully!\n")

# Print dataset information (uncomment after creating variables above):
# cat("Dataset Dimensions:\n")
# cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
# cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
# cat("Order Items:", nrow(order_items), "rows x", ncol(order_items), "columns\n")
# cat("Products:", nrow(products), "rows x", ncol(products), "columns\n")
# cat("Suppliers:", nrow(suppliers), "rows x", ncol(suppliers), "columns\n")

# Show first 3 rows of each dataset (uncomment after loading data):
# head(customers, 3)
# head(orders, 3)
# head(order_items, 3)
# head(products, 3)
# head(suppliers, 3)

✅ Working directory set to: /workspaces/Carsons-stuff/data 


Rows: 100 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Name, Email, City
dbl  (1): CustomerID
date (1): Registration_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 250 Columns: 4


── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl  (3): OrderID, CustomerID, Total_Amount
date (1): Order_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 400 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (4): OrderID, ProductID, Quantity, Unit_Price

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 50 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): Product_Name, Category
dbl (2): ProductID, Supplier_ID

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 10 Columns: 3
── Colu

✅ Datasets imported successfully!


## Part 2: Basic Joins

**Tasks:**
1. **Inner Join:** Create `customer_orders_inner` by joining customers and orders
2. **Left Join:** Create `customer_orders_left` to show all customers  
3. **Right Join:** Create `customer_orders_right` to show all orders
4. **Full Join:** Create `customer_orders_full` for complete data view

Use the exact variable names specified above for autograding. Analyze row counts and explain results.

In [4]:
# Your code here:
# Create an inner join between customers and orders
inner_joined_data <- inner_join(customers, orders, by = "CustomerID")
print("Inner Join (matching CustomerIDs only):")
print(inner_joined_data)
# Analyze the results and compare row counts
#Notice Customers 1-5 all have orders and some of the OrderID values may be missing because there isnt an order ID that matches that customer.


[1] "Inner Join (matching CustomerIDs only):"


# A tibble: 200 × 8
   CustomerID Name       Email        City  Registration_Date OrderID Order_Date
        <dbl> <chr>      <chr>        <chr> <date>              <dbl> <date>    
 1          1 Customer 1 customer1@e… Phoe… 2020-10-03             87 2023-03-28
 2          1 Customer 1 customer1@e… Phoe… 2020-10-03            214 2023-09-12
 3          2 Customer 2 customer2@e… Los … 2020-06-02            173 2024-02-25
 4          2 Customer 2 customer2@e… Los … 2020-06-02            190 2023-04-19
 5          3 Customer 3 customer3@e… Chic… 2021-04-20             29 2023-03-07
 6          3 Customer 3 customer3@e… Chic… 2021-04-20            146 2023-04-10
 7          4 Customer 4 customer4@e… Hous… 2022-06-16              7 2023-05-01
 8          4 Customer 4 customer4@e… Hous… 2022-06-16             61 2024-01-19
 9          5 Customer 5 customer5@e… Phoe… 2023-08-22             15 2023-12-27
10          5 Customer 5 customer5@e… Phoe… 2023-08-22            150 2023-03-28
# ℹ 190 

In [5]:
# Your code here:
# Create a left join to keep all customers
left_joined_data <- left_join(customers, orders, by = "CustomerID")
print("Left Join (all customers, matching orders):")
print(left_joined_data)
# Count how many customers have no orders
#Notice 0 people dont have an order. Everyone has an order. It would show NA for order date and ID if they didnt have an order.


[1] "Left Join (all customers, matching orders):"


# A tibble: 200 × 8
   CustomerID Name       Email        City  Registration_Date OrderID Order_Date
        <dbl> <chr>      <chr>        <chr> <date>              <dbl> <date>    
 1          1 Customer 1 customer1@e… Phoe… 2020-10-03             87 2023-03-28
 2          1 Customer 1 customer1@e… Phoe… 2020-10-03            214 2023-09-12
 3          2 Customer 2 customer2@e… Los … 2020-06-02            173 2024-02-25
 4          2 Customer 2 customer2@e… Los … 2020-06-02            190 2023-04-19
 5          3 Customer 3 customer3@e… Chic… 2021-04-20             29 2023-03-07
 6          3 Customer 3 customer3@e… Chic… 2021-04-20            146 2023-04-10
 7          4 Customer 4 customer4@e… Hous… 2022-06-16              7 2023-05-01
 8          4 Customer 4 customer4@e… Hous… 2022-06-16             61 2024-01-19
 9          5 Customer 5 customer5@e… Phoe… 2023-08-22             15 2023-12-27
10          5 Customer 5 customer5@e… Phoe… 2023-08-22            150 2023-03-28
# ℹ 190 

In [6]:
# Your code here:
# Create a right join to keep all orders
right_joined_data <- right_join(customers, orders, by = "CustomerID")
print("Right Join (all orders, matching customers):")
print(right_joined_data)
# Check for orders with invalid customer IDs
Everyone has an OrderID.


ERROR: Error in parse(text = input): <text>:7:10: unexpected symbol
6: # Check for orders with invalid customer IDs
7: Everyone has
            ^


In [8]:
# Part 2: Basic Joins

# Part 2.1: Create customer_orders using inner_join()
customer_orders <- customers %>%
  inner_join(orders, by = "CustomerID")

# Part 2.2: Create customer_orders_left using left_join() 
customer_orders_left <- customers %>%
  left_join(orders, by = "CustomerID")

# Part 2.3: Create customer_orders_right using right_join()
customer_orders_right <- customers %>%
  right_join(orders, by = "CustomerID")

# Part 2.4: Create customer_orders_full using full_join()
customer_orders_full <- customers %>%
  full_join(orders, by = "CustomerID")

# Analysis and comparison
cat("\n=== JOIN ANALYSIS ===\n\n")

cat("Original Dataset Sizes:\n")
cat("Customers:", nrow(customers), "rows\n")
cat("Orders:", nrow(orders), "rows\n\n")

cat("Join Results:\n")
cat("Inner Join (customer_orders):", nrow(customer_orders), "rows\n")
cat("Left Join (customer_orders_left):", nrow(customer_orders_left), "rows\n")
cat("Right Join (customer_orders_right):", nrow(customer_orders_right), "rows\n")
cat("Full Join (customer_orders_full):", nrow(customer_orders_full), "rows\n\n")

# Analysis of unmatched records
cat("=== UNMATCHED RECORDS ANALYSIS ===\n\n")

# Customers without orders
customers_without_orders <- customer_orders_left %>%
  filter(is.na(OrderID)) %>%
  nrow()
cat("Customers WITHOUT orders:", customers_without_orders, "\n")

# Orders without matching customers
orders_without_customers <- customer_orders_right %>%
  filter(is.na(Name)) %>%
  nrow()
cat("Orders WITHOUT matching customers:", orders_without_customers, "\n")

# Total unmatched in full join
total_unmatched <- customer_orders_full %>%
  filter(is.na(OrderID) | is.na(Name)) %>%
  nrow()
cat("Total unmatched records (full join):", total_unmatched, "\n\n")

# Summary
cat("=== INTERPRETATION ===\n")
cat("• Inner Join: Only matched customer-order pairs\n")
cat("• Left Join: All customers + their orders (NAs for customers without orders)\n")
cat("• Right Join: All orders + customer info (NAs for orders without customers)\n")
cat("• Full Join: Everything from both tables (NAs where no match exists)\n")


=== JOIN ANALYSIS ===

Original Dataset Sizes:
Customers: 100 rows
Orders: 250 rows

Join Results:
Inner Join (customer_orders): 200 rows
Left Join (customer_orders_left): 200 rows
Right Join (customer_orders_right): 250 rows
Full Join (customer_orders_full): 250 rows

=== UNMATCHED RECORDS ANALYSIS ===

Customers WITHOUT orders: 0 
Orders WITHOUT matching customers: 50 
Total unmatched records (full join): 50 

=== INTERPRETATION ===
• Inner Join: Only matched customer-order pairs
• Left Join: All customers + their orders (NAs for customers without orders)
• Right Join: All orders + customer info (NAs for orders without customers)
• Full Join: Everything from both tables (NAs where no match exists)


In [9]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
customer_orders <- customers %>%
  inner_join(orders, by = "CustomerID")

# Required output for autograding:
cat("Inner Join Results:\n")
cat("Customers:", nrow(customers), "\n")
cat("Orders:", nrow(orders), "\n")
cat("Inner Join Result:", nrow(customer_orders), "\n")# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
customer_orders <- customers %>%
  inner_join(orders, by = "CustomerID")

# Required output for autograding:
cat("Inner Join Results:\n")
cat("Customers:", nrow(customers), "\n")
cat("Orders:", nrow(orders), "\n")
cat("Inner Join Result:", nrow(customer_orders), "\n")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 
Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [10]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
customer_orders_left <- customers %>%
  left_join(orders, by = "CustomerID")

# Required output for autograding:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:\n")
cat("Total rows:", nrow(customer_orders_left), "\n")
cat("Customers without orders:", customers_without_orders, "\n")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [11]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
customer_orders <- customers %>%
  inner_join(orders, by = "CustomerID")

# Expected output:
cat("Inner Join Results:\n")
cat("Customers:", nrow(customers), "\n")
cat("Orders:", nrow(orders), "\n")
cat("Inner Join Result:", nrow(customer_orders), "\n")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [12]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
customer_orders_left <- customers %>%
  left_join(orders, by = "CustomerID")

# Expected output:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:\n")
cat("Total rows:", nrow(customer_orders_left), "\n")
cat("Customers without orders:", customers_without_orders, "\n")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [13]:
# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
customer_orders_right <- customers %>%
  right_join(orders, by = "CustomerID")

# Expected output:
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:\n")
cat("Total rows:", nrow(customer_orders_right), "\n")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "\n")

Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 


In [14]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
customer_orders_full <- customers %>%
  full_join(orders, by = "CustomerID")

# Expected output:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:\n")
cat("Total rows:", nrow(customer_orders_full), "\n")
cat("Customers without orders:", customers_only, "\n")
cat("Orders without valid customers:", orders_only, "\n")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [15]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
customer_orders_full <- customers %>%
  full_join(orders, by = "CustomerID")

# Required output for autograding:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:\n")
cat("Total rows:", nrow(customer_orders_full), "\n")
cat("Customers without orders:", customers_only, "\n")
cat("Orders without valid customers:", orders_only, "\n")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


## Part 3: Multi-Table Joins

**Tasks:**
1. Create a comprehensive dataset by joining `orders`, `customers`, and `order_items`
2. Extend this dataset by adding `products` and `suppliers` information
3. Create a complete supply chain view showing the full customer-to-supplier relationship

Build your joins step by step and examine the results at each stage.

## Part 3: Multi-Table Joins

**Requirements for Autograding:**
You must create these EXACT variable names in order:

1. `orders_items` - Join orders and order_items  
2. `orders_customers_items` - Add customers to above
3. `complete_order_data` - Add products information
4. `complete_data` - Add suppliers for complete supply chain view

**Expected Outputs:**
Print the number of rows for each step to show the join progression.

In [16]:
# Step 1: REQUIRED variable name: orders_items
# Your code here:
orders_items <- orders %>%
  inner_join(order_items, by = "OrderID")

# Required output for autograding:
cat("Step 1 - Orders with Items:", nrow(orders_items), "rows\n")

Step 1 - Orders with Items: 400 rows


In [17]:
# Step 2: REQUIRED variable name: orders_customers_items  
# Your code here:
orders_customers_items <- orders_items %>%
  inner_join(customers, by = "CustomerID")

# Required output for autograding:
cat("Step 2 - Add Customers:", nrow(orders_customers_items), "rows\n")

Step 2 - Add Customers: 310 rows


In [18]:
# Step 3: REQUIRED variable name: complete_order_data
# Your code here:
complete_order_data <- orders_customers_items %>%
  inner_join(products, by = "ProductID")

# Required output for autograding:
cat("Step 3 - Add Products:", nrow(complete_order_data), "rows\n")

Step 3 - Add Products: 310 rows


In [19]:
# Step 4: REQUIRED variable name: complete_data
# Your code here:
complete_data <- complete_order_data %>%
  inner_join(suppliers, by = "Supplier_ID")

# Required output for autograding:
cat("Step 4 - Complete Supply Chain:", nrow(complete_data), "rows\n")

Step 4 - Complete Supply Chain: 310 rows


## Part 4: Data Quality Analysis

**Tasks:**
1. Use `anti_join()` to find unmatched records between tables
2. Use `semi_join()` to find matched records
3. Check for duplicate keys and analyze their impact on joins
4. Identify and document any data quality issues

This analysis helps understand data integrity and potential issues in your datasets.

In [20]:
# Part 4: Data Quality Analysis

# 1. customers_no_orders - Use anti_join() to find customers with no orders
customers_no_orders <- customers %>%
  anti_join(orders, by = "CustomerID")

# 2. orphaned_orders - Use anti_join() to find orders without customers  
orphaned_orders <- orders %>%
  anti_join(customers, by = "CustomerID")

# 3. products_never_ordered - Use anti_join() to find unordered products
products_never_ordered <- products %>%
  anti_join(order_items, by = "ProductID")

# 4. active_customers - Use semi_join() to find customers with orders
active_customers <- customers %>%
  semi_join(orders, by = "CustomerID")

# Expected Outputs - Print the count of records for each
cat("Data Quality Analysis:\n")
cat("Customers with no orders:", nrow(customers_no_orders), "\n")
cat("Orphaned orders (no customer):", nrow(orphaned_orders), "\n")
cat("Products never ordered:", nrow(products_never_ordered), "\n")
cat("Active customers (with orders):", nrow(active_customers), "\n")

Data Quality Analysis:
Customers with no orders: 0 
Orphaned orders (no customer): 50 
Products never ordered: 0 
Active customers (with orders): 100 


In [21]:
# REQUIRED variable name: customers_no_orders
# Your code here:
customers_no_orders <- customers %>%
  anti_join(orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

Customers who never placed an order: 0 


In [22]:
# REQUIRED variable name: orphaned_orders
# Your code here:
orphaned_orders <- orders %>%
  anti_join(customers, by = "CustomerID")

# Required output for autograding:
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

Orders without corresponding customers: 50 


In [23]:
# REQUIRED variable name: products_never_ordered
# Your code here:
products_never_ordered <- products %>%
  anti_join(order_items, by = "ProductID")

# Required output for autograding:
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

Products that were never ordered: 0 


In [24]:
# REQUIRED variable name: active_customers
# Your code here:
active_customers <- customers %>%
  semi_join(orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who placed at least one order: 100 


In [25]:
# Data integrity checks - count duplicate keys
# Your code here: Calculate and print duplicate counts for each dataset

# Required outputs for autograding (uncomment after creating the calculations above):
duplicate_customers <- sum(duplicated(customers$CustomerID))
duplicate_orders <- sum(duplicated(orders$OrderID))
duplicate_products <- sum(duplicated(products$ProductID))
duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

cat("Data Quality Summary:\n")
cat("Duplicate customer IDs:", duplicate_customers, "\n")
cat("Duplicate order IDs:", duplicate_orders, "\n")
cat("Duplicate product IDs:", duplicate_products, "\n")
cat("Duplicate supplier IDs:", duplicate_suppliers, "\n")

Data Quality Summary:
Duplicate customer IDs: 0 
Duplicate order IDs: 0 
Duplicate product IDs: 0 
Duplicate supplier IDs: 0 


## Part 5: Business Analysis with Joined Data

**Tasks:**
1. Calculate customer lifetime value metrics (total spent, number of orders, average order value)
2. Analyze product performance (total quantity sold, revenue generated, order frequency)
3. Evaluate supplier performance (total revenue, number of products supplied)
4. Create a regional analysis by customer city

Use your joined datasets to generate meaningful business insights.

In [26]:
# --- Part 5: Business Analysis with Joined Data ---

# Customer metrics
customer_metrics <- complete_data %>%
  group_by(CustomerID, Name) %>%
  summarise(
    Total_Spent = sum(Total_Amount, na.rm = TRUE),
    Order_Count = n_distinct(OrderID),
    Avg_Order_Value = Total_Spent / Order_Count,
    .groups = "drop"
  )

# Product metrics
product_metrics <- complete_data %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Total_Quantity = sum(Quantity, na.rm = TRUE),
    Order_Frequency = n_distinct(OrderID),
    .groups = "drop"
  )

# Supplier metrics
supplier_metrics <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Product_Count = n_distinct(ProductID),
    .groups = "drop"
  )

# Regional analysis
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  )

# --- Display results ---
cat("Business Analysis Summary:\n\n")

cat("Customer Metrics - Top 5 Customers:\n")
print(head(customer_metrics %>% arrange(desc(Total_Spent)), 5))

cat("\nProduct Metrics - Top 5 Products:\n")
print(head(product_metrics %>% arrange(desc(Total_Revenue)), 5))

cat("\nSupplier Metrics:\n")
print(supplier_metrics)

cat("\nRegional Analysis:\n")
print(regional_analysis %>% arrange(desc(Total_Sales)))


Business Analysis Summary:



Customer Metrics - Top 5 Customers:
# A tibble: 5 × 5
  CustomerID Name        Total_Spent Order_Count Avg_Order_Value
       <dbl> <chr>             <dbl>       <int>           <dbl>
1         88 Customer 88      12115.           2           6058.
2         63 Customer 63      11653.           2           5827.
3         82 Customer 82       9371.           2           4686.
4         49 Customer 49       9021.           2           4510.
5         94 Customer 94       8731.           2           4366.

Product Metrics - Top 5 Products:
# A tibble: 5 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        47 Product 47          11763.             35               8
2        26 Product 26          10854.             42              13
3        43 Product 43          10584.             36              10
4        13 Product 13           9119.             30               8
5        18 Produc

In [27]:
# REQUIRED variable name: customer_metrics
# Your code here: Calculate customer lifetime value metrics
# Required columns: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
# 1. customer_metrics - Customer analysis with Total_Spent, Order_Count, Avg_Order_Value
customer_metrics <- complete_data %>%
  group_by(CustomerID, Name) %>%
  summarise(
    Total_Spent = sum(Total_Amount),
    Order_Count = n_distinct(OrderID),
    Avg_Order_Value = Total_Spent / Order_Count,
    .groups = "drop"
  )

# 2. product_metrics - Product analysis with Total_Revenue, Total_Quantity, Order_Frequency
product_metrics <- complete_data %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price),
    Total_Quantity = sum(Quantity),
    Order_Frequency = n_distinct(OrderID),
    .groups = "drop"
  )

# 3. supplier_metrics - Supplier analysis with Product_Count by supplier
supplier_metrics <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Product_Count = n_distinct(ProductID),
    .groups = "drop"
  )

# 4. regional_analysis - Sales analysis by customer city
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price),
    Customer_Count = n_distinct(CustomerID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  )

# Display results
cat("Business Analysis Summary:\n\n")
cat("Customer Metrics - Top 5 Customers:\n")
print(head(customer_metrics %>% arrange(desc(Total_Spent)), 5))

cat("\nProduct Metrics - Top 5 Products:\n")
print(head(product_metrics %>% arrange(desc(Total_Revenue)), 5))

cat("\nSupplier Metrics:\n")
print(supplier_metrics)

cat("\nRegional Analysis:\n")
print(regional_analysis %>% arrange(desc(Total_Sales)))

# Required output for autograding:
cat("Customer Analysis Summary:
")
cat("Total customers analyzed:", nrow(customer_metrics), "
")
cat("Top customer total spent: $", round(max(customer_metrics$Total_Spent, na.rm = TRUE), 2), "
")
print(head(customer_metrics[order(-customer_metrics$Total_Spent), ], 3))

Business Analysis Summary:

Customer Metrics - Top 5 Customers:
# A tibble: 5 × 5
  CustomerID Name        Total_Spent Order_Count Avg_Order_Value
       <dbl> <chr>             <dbl>       <int>           <dbl>
1         88 Customer 88      12115.           2           6058.
2         63 Customer 63      11653.           2           5827.
3         82 Customer 82       9371.           2           4686.
4         49 Customer 49       9021.           2           4510.
5         94 Customer 94       8731.           2           4366.

Product Metrics - Top 5 Products:
# A tibble: 5 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        47 Product 47          11763.             35               8
2        26 Product 26          10854.             42              13
3        43 Product 43          10584.             36              10
4        13 Product 13           9119.             30       

In [28]:
# REQUIRED variable name: product_metrics  
# Your code here: Analyze product performance
# Required columns: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
product_metrics <- complete_data %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),  # total sales for product
    Total_Quantity = sum(Quantity, na.rm = TRUE),              # total units sold
    Order_Frequency = n_distinct(OrderID),                     # how many orders included this product
    .groups = "drop"
  )

# Required output for autograding:
cat("Product Analysis Summary:
")
cat("Total products analyzed:", nrow(product_metrics), "
")
cat("Top product revenue: $", round(max(product_metrics$Total_Revenue, na.rm = TRUE), 2), "
")
print(head(product_metrics[order(-product_metrics$Total_Revenue), ], 3))

Product Analysis Summary:
Total products analyzed: 50 
Top product revenue: $ 11763.16 
# A tibble: 3 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        47 Product 47          11763.             35               8
2        26 Product 26          10854.             42              13
3        43 Product 43          10584.             36              10


In [29]:
# REQUIRED variable name: supplier_metrics
# Your code here: Evaluate supplier performance  
# Required columns: Supplier_ID, Supplier_Name, Product_Count
supplier_metrics <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Product_Count = n_distinct(ProductID),   # how many unique products each supplier provides
    .groups = "drop"
  )

# Required output for autograding:
cat("Supplier Analysis Summary:
")
cat("Total suppliers analyzed:", nrow(supplier_metrics), "
")
cat("Max products per supplier:", max(supplier_metrics$Product_Count, na.rm = TRUE), "
")
print(supplier_metrics[order(-supplier_metrics$Product_Count), ])

Supplier Analysis Summary:
Total suppliers analyzed: 10 
Max products per supplier: 10 
# A tibble: 10 × 3
   Supplier_ID Supplier_Name Product_Count
         <dbl> <chr>                 <int>
 1           7 Supplier 7               10
 2           4 Supplier 4                6
 3          10 Supplier 10               6
 4           1 Supplier 1                5
 5           3 Supplier 3                5
 6           5 Supplier 5                5
 7           6 Supplier 6                5
 8           2 Supplier 2                3
 9           8 Supplier 8                3
10           9 Supplier 9                2


In [30]:
# REQUIRED variable name: regional_analysis
# Your code here: Create regional analysis by customer city
# Required columns: City, Total_Sales, Customer_Count, Avg_Customer_Value
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),   # total sales per city
    Customer_Count = n_distinct(CustomerID),                  # number of unique customers in the city
    Avg_Customer_Value = Total_Sales / Customer_Count,        # average revenue per customer
    .groups = "drop"
  )

# Required output for autograding:
cat("Regional Analysis Summary:
")
cat("Total cities analyzed:", nrow(regional_analysis), "
")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "
")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 5 
Highest city sales: $ 72277.52 
# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             23              3143.
2 Phoenix          49333.             19              2596.
3 Los Angeles      48506.             18              2695.
4 Houston          38143.             18              2119.
5 New York         36417.             16              2276.


In [31]:
# Your code here:
# Analyze product performance metrics
product_metrics <- complete_data %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),  # total revenue per product
    Total_Quantity = sum(Quantity, na.rm = TRUE),              # total units sold
    Order_Frequency = n_distinct(OrderID),                     # number of unique orders
    .groups = "drop"
  )


In [32]:
# Your code here:
# Evaluate supplier performance
supplier_metrics <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Product_Count = n_distinct(ProductID),  # number of unique products per supplier
    .groups = "drop"
  )


In [33]:
# Your code here:
# Create regional analysis by customer city
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),   # total sales per city
    Customer_Count = n_distinct(CustomerID),                  # number of unique customers per city
    Avg_Customer_Value = Total_Sales / Customer_Count,        # average sales per customer
    .groups = "drop"
  )


In [34]:
# Supplier Performance Analysis
supplier_metrics <- complete_data %>%
  filter(!is.na(Supplier_Name)) %>%
  group_by(Supplier_ID, Supplier_Name, Country) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))

cat("Supplier Performance:\n")
print(supplier_metrics)

Supplier Performance:
# A tibble: 10 × 6
   Supplier_ID Supplier_Name Country     Total_Revenue Products_Supplied
         <dbl> <chr>         <chr>               <dbl>             <int>
 1           5 Supplier 5    Japan              43434.                 5
 2           7 Supplier 7    Germany            35416.                10
 3          10 Supplier 10   Japan              31875.                 6
 4           1 Supplier 1    USA                29706.                 5
 5           3 Supplier 3    Japan              25989.                 5
 6           4 Supplier 4    South Korea        25341.                 6
 7           6 Supplier 6    Japan              21003.                 5
 8           9 Supplier 9    Japan              14845.                 2
 9           2 Supplier 2    USA                10263.                 3
10           8 Supplier 8    Germany             6805.                 3
# ℹ 1 more variable: Orders_Involved <int>


In [35]:
# Regional Analysis by Customer City
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Sales))

cat("Regional Sales Analysis:\n")
print(regional_analysis)

Regional Sales Analysis:
# A tibble: 5 × 5
  City        Total_Sales Customer_Count Order_Count Avg_Customer_Value
  <chr>             <dbl>          <int>       <int>              <dbl>
1 Chicago          72278.             23          36              3143.
2 Phoenix          49333.             19          32              2596.
3 Los Angeles      48506.             18          32              2695.
4 Houston          38143.             18          30              2119.
5 New York         36417.             16          26              2276.


## Part 6: Complex Business Questions

**Tasks:**
Answer the following business questions using your joined datasets:

1. **Customer Segmentation:** Identify your top 10% of customers by total spending
2. **Product Recommendations:** Which products are frequently bought together?
3. **Supplier Dependency:** Which suppliers are most critical to the business?
4. **Market Expansion:** Which cities have high customer counts but low average order values?

Provide data-driven answers with supporting analysis.

## Part 6: Complex Business Questions

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `top_customers` - Top 10% of customers by spending (use quantile function)
2. `product_combinations` - Products frequently bought together  
3. `critical_suppliers` - Suppliers ranked by importance
4. `market_expansion` - Cities with expansion opportunities

**Expected Outputs:**
Each analysis should include specific metrics and rankings as shown in the required outputs below.

In [36]:
# --- Load required packages ---
library(dplyr)

# --- Step 1: Assume you have a dataset of orders (replace with your data) ---
# Example sample data frame
complete_data <- data.frame(
  CustomerID = c(1,1,2,2,3,3,4,5,5,5),
  Name = c("Alice","Alice","Bob","Bob","Carol","Carol","Dan","Eve","Eve","Eve"),
  OrderID = c(101,102,103,104,105,106,107,108,109,110),
  Total_Amount = c(200,150,300,250,400,350,120,500,600,700)
)

# --- Step 2: Create `customer_metrics` dataset ---
# Summarize total spending per customer
customer_metrics <- complete_data %>%
  group_by(CustomerID, Name) %>%
  summarise(
    Total_Spent = sum(Total_Amount, na.rm = TRUE),
    Order_Count = n_distinct(OrderID),
    Avg_Order_Value = Total_Spent / Order_Count,
    .groups = "drop"
  )

# --- Step 3: Identify Top 10% of Customers by Spending ---
# Calculate the 90th percentile spending threshold
top_10_percent_threshold <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)

# Filter customers who are above or equal to this threshold
top_customers <- customer_metrics %>%
  filter(Total_Spent >= top_10_percent_threshold)

# --- Step 4: Required Output for Autograding ---
cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_10_percent_threshold, 2), "\n")
cat("Number of top customers:", nrow(top_customers), "\n")

# --- Step 5: Optional — View top customers ---
head(top_customers, 10)



Customer Segmentation Results:
Top 10% spending threshold: $ 1380 
Number of top customers: 1 


CustomerID Name Total_Spent Order_Count Avg_Order_Value
1 5          Eve  1800        3           600

In [37]:
library(dplyr)
library(tidyr)

# --- Example dataset (replace with your own complete_data) ---
complete_data <- data.frame(
  OrderID = c(1, 1, 1, 2, 2, 3, 3, 3, 4, 5, 5),
  ProductID = c(10, 11, 12, 10, 13, 14, 15, 10, 14, 16, 17),
  Product_Name = c(
    "Keyboard", "Mouse", "Monitor",
    "Keyboard", "Webcam",
    "Chair", "Desk", "Keyboard",
    "Chair", "Headphones", "Microphone"
  )
)

# --- Step 1: Keep only orders with more than 1 product ---
multi_item_orders <- complete_data %>%
  group_by(OrderID) %>%
  filter(n_distinct(ProductID) > 1) %>%
  select(OrderID, ProductID, Product_Name)

# --- Step 2: Generate product pairs for each order ---
product_combinations <- multi_item_orders %>%
  group_by(OrderID) %>%
  summarise(
    pairs = list({
      items <- sort(unique(Product_Name))
      if (length(items) > 1) {
        as.data.frame(t(combn(items, 2)), stringsAsFactors = FALSE)
      } else {
        NULL
      }
    }),
    .groups = "drop"
  ) %>%
  unnest(cols = c(pairs)) %>%
  rename(Product1 = V1, Product2 = V2)

# --- Step 3: Count frequency of each combination ---
product_combinations <- product_combinations %>%
  group_by(Product1, Product2) %>%
  summarise(n = n(), .groups = "drop") %>%
  arrange(desc(n))

# --- Step 4: Required output for autograding ---
cat("Product Combination Analysis:\n")
cat("Total product combinations found:", nrow(product_combinations), "\n")
if (nrow(product_combinations) > 0) {
  cat("Most frequent combination appears", max(product_combinations$n, na.rm = TRUE), "times\n")
}

# --- Step 5: Optional — see top combinations ---
head(product_combinations, 10)



Product Combination Analysis:
Total product combinations found: 8 
Most frequent combination appears 1 times


Product1   Product2   n
1 Chair      Desk       1
2 Chair      Keyboard   1
3 Desk       Keyboard   1
4 Headphones Microphone 1
5 Keyboard   Monitor    1
6 Keyboard   Mouse      1
7 Keyboard   Webcam     1
8 Monitor    Mouse      1

In [38]:
library(dplyr)

# --- Ensure complete_data exists (for testing only) ---
# Remove this block if your dataset is already loaded
complete_data <- data.frame(
  Supplier_ID = c(1, 1, 2, 2, 3, 3, 3, 4),
  Supplier_Name = c("TechSource", "TechSource", "CompuWorld", "CompuWorld", "OfficePro", "OfficePro", "OfficePro", "ElectroHub"),
  ProductID = c(10, 11, 12, 13, 14, 15, 16, 17),
  Quantity = c(100, 150, 120, 80, 200, 160, 90, 50),
  Unit_Price = c(10, 20, 15, 25, 12, 18, 10, 40)
)

# --- REQUIRED variable name: critical_suppliers ---
# Analyze supplier performance: revenue contribution + product diversity

critical_suppliers <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Product_Count = n_distinct(ProductID),
    .groups = "drop"
  ) %>%
  # Combine both metrics into a "Criticality_Score"
  mutate(
    Revenue_Share = Total_Revenue / sum(Total_Revenue),
    Diversity_Share = Product_Count / sum(Product_Count),
    Criticality_Score = (Revenue_Share * 0.7) + (Diversity_Share * 0.3)
  ) %>%
  arrange(desc(Criticality_Score))

# --- Required output for autograding ---
cat("Critical Suppliers Analysis:\n")
cat("Total suppliers analyzed:", nrow(critical_suppliers), "\n")
print("Top 3 most critical suppliers:")
print(head(critical_suppliers, 3))


Critical Suppliers Analysis:
Total suppliers analyzed: 4 
[1] "Top 3 most critical suppliers:"
# A tibble: 3 × 7
  Supplier_ID Supplier_Name Total_Revenue Product_Count Revenue_Share
        <dbl> <chr>                 <dbl>         <int>         <dbl>
1           3 OfficePro              6180             3         0.387
2           1 TechSource             4000             2         0.250
3           2 CompuWorld             3800             2         0.238
# ℹ 2 more variables: Diversity_Share <dbl>, Criticality_Score <dbl>


In [39]:
library(dplyr)

# --- Example dataset (remove this section if complete_data is already loaded) ---
# Assuming complete_data includes CustomerID, City, OrderID, and Total_Amount
complete_data <- data.frame(
  CustomerID = rep(1:10, each = 2),
  City = c("Phoenix", "Phoenix", "Chicago", "Chicago", "Houston", "Houston", "LA", "LA", "NYC", "NYC",
           "Phoenix", "Chicago", "Houston", "LA", "NYC", "Chicago", "Houston", "Phoenix", "LA", "NYC"),
  OrderID = 1:20,
  Total_Amount = c(100, 200, 300, 120, 250, 700, 800, 400, 500, 900,
                   150, 100, 600, 700, 300, 500, 450, 800, 200, 400)
)

# --- REQUIRED variable name: market_expansion ---
# Identify cities with multiple customers and high variation in order values

market_expansion <- complete_data %>%
  group_by(City, CustomerID) %>%
  summarise(Avg_Order_Value = mean(Total_Amount, na.rm = TRUE), .groups = "drop") %>%
  group_by(City) %>%
  summarise(
    Customer_Count = n_distinct(CustomerID),
    Avg_City_Order_Value = mean(Avg_Order_Value, na.rm = TRUE),
    Order_Value_SD = sd(Avg_Order_Value, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  filter(Customer_Count > 1 & !is.na(Order_Value_SD)) %>%
  mutate(
    Opportunity_Index = Order_Value_SD * log(Customer_Count + 1)
  ) %>%
  arrange(desc(Opportunity_Index))

# --- Required output for autograding ---
cat("Market Expansion Analysis:\n")
cat("Cities evaluated for expansion:", nrow(market_expansion), "\n")
print("Top expansion opportunities:")
print(head(market_expansion, 5))


Market Expansion Analysis:
Cities evaluated for expansion: 5 
[1] "Top expansion opportunities:"


# A tibble: 5 × 5
  City    Customer_Count Avg_City_Order_Value Order_Value_SD Opportunity_Index
  <chr>            <int>                <dbl>          <dbl>             <dbl>
1 Phoenix              3                 367.          375.               520.
2 LA                   3                 500           265.               367.
3 NYC                  3                 467.          208.               289.
4 Chicago              3                 270           207.               286.
5 Houston              3                 508.           80.4              111.


## Part 7: Summary and Insights

**Requirements for Autograding:**
Complete the analysis summary below with specific findings from your work. Your responses will be evaluated for completeness and accuracy.

**Grading Criteria:**
- Specific metrics and numbers from your analysis
- Clear identification of data quality issues  
- Actionable business recommendations
- Technical insights about join performance and appropriateness

## Analysis Summary

### Key Findings from Join Operations:
- **Join Efficiency:** The inner join is 80%. The left join 100%. Right join 80%. Full join is 80%.
- **Data Relationships:** There is a one to many relationship because customers can have multiple orders, everyone has orders, and some orders were missing valid customerIDs.
- **Multi-table Joins:** The inner join shows strict matches only while the left join shows all customers. the right join shows all records, and the full join shows all of the records from both sides.

### Data Quality Issues Discovered:
- **Orphaned Records:** There were 50 orphaned orders.
- **Missing Data:** There are 50 orders with the customers missing from them. There are no customers with missing orders or no orders.
- **Referential Integrity:** There are no foreign key violations. 

### Business Insights:
- **Top Customer:** Customer 88      12115
- **Best Product:** Product 47          11763 
- **Regional Performance:** They sold the most in Chicago, Phoenix and LA.
- **Supplier Analysis:** A large majority (10) of their products comes from supplier 7.

### Strategic Recommendations:
- Find the demographic of Customer 88 and target that audience.
- More focus can be shifted toward producing and selling product 47 or making products with similar features to product 47.
- They can target Chicago, Phoenix, and LA.

### Technical Learnings:
- **Inner vs Left Joins:** The left join is used when you want all the data from the left set and the matching data between the left and right set. A middle join is used when you want to see the data that just matches between the two sets.
- **Multi-table Strategy:** Only join clean data sets with little to no missing values that have already been sorted using a left join for more complex data sets so some structure still remains. 
- **Performance Considerations:** A full join is good for small data sets when seeking general information. An inner join is useful for large data sets when you're trying to find very specific information. Left and right joins are good for additional information to be seen.

In [40]:
# Optional: Save your analysis results (not required for autograding)
# Your code here: Save key datasets to CSV files if desired


# Final summary output for autograding verification:
cat("\n=== HOMEWORK 6 COMPLETION SUMMARY ===\n")
cat("✓ Data Import: 5 datasets loaded\n")
cat("✓ Basic Joins: 4 join types completed\n") 
cat("✓ Multi-table Joins: 4-step progression completed\n")
cat("✓ Data Quality: Anti-joins and semi-joins performed\n")
cat("✓ Business Analysis: Customer, product, supplier, and regional analysis completed\n")
cat("✓ Complex Questions: Advanced business scenarios analyzed\n")
cat("✓ Summary: Comprehensive insights documented\n")
cat("\nAll required variables created for autograding verification.\n")


=== HOMEWORK 6 COMPLETION SUMMARY ===
✓ Data Import: 5 datasets loaded
✓ Basic Joins: 4 join types completed
✓ Multi-table Joins: 4-step progression completed
✓ Data Quality: Anti-joins and semi-joins performed
✓ Business Analysis: Customer, product, supplier, and regional analysis completed
✓ Complex Questions: Advanced business scenarios analyzed
✓ Summary: Comprehensive insights documented

All required variables created for autograding verification.
